### ISSUE TITLE
How exactly do 'direct-space', 'coulomb', and 'exact' options in AbsoluteAlchemicalFactory treat electrostatics?

### ISSUE THREAD

**Issue by JIMonroe** [choderalab/openmmtools#376]
I'm looking to perform some alchemical free energy calculations and recently started using OpenMM to do this.  You've all done some nice work here, so I don't want to re-invent the wheel, but I initially set out to try and make these calculations work without openmmtools.alchemy, mainly to deepen my own understanding of how this all works.  

Just to clearly define the problem, I want to take a particular molecule and decouple it from the rest of the system, leaving all direct-space intramolecular interactions on.  From what I understand, this is what GROMACS achieves (someone please correct me here if necessary), and so I've been checking potential energies at different states against calculations from GROMACS.  

This is relatively straight-forward in terms of sterics, as there is no reciprocal space term to deal with.  However, electrostatics is a different story.  Ideally, the coupled state includes all direct and reciprocal space interactions while the uncoupled state includes direct and reciprocal space interactions for the system minus the alchemical molecule and the direct-space interactions for the molecule with itself.  I think I figured out a way to do this with OpenMM, and it mostly matches with GROMACS, but it doesn't seem to match any of the options for handling electrostatics provided by AbsoluteAlchemicalFactory.  In my approach, I use the following scheme:

- NonbondedForce using PME with all electrostatics (except 1-4 terms and with all alchemical/alchemical terms set to zero with exceptions), and all non-alchemical/non-alchemical sterics (LJ)

- CustomNonbondedForce using CutoffPeriodic implementing a soft-core potential for alchemical/non-alchemical sterics

- NonbondedForce using PME with ONLY exceptions (zero for all particle charges and epsilons), including 1-4 terms and other alchemical/alchemical non-bonded interactions

Is this doing what I think it is?  And for the options you have in AbsoluteAlchemicalFactory, what are the exact end states (coupled molecule with lambda of say 0, to uncoupled molecule with lambda of 1)?  In terms of the potential energy function?

In my tests, I have set up the alchemical system with the following code:

```
alchFactory = alchemy.AbsoluteAlchemicalFactory(consistent_exceptions=False, 
                                                alchemical_pme_treatment='coulomb',
                                                disable_alchemical_dispersion_correction=False,
                                                split_alchemical_forces=True
                                               )

alchemicalRegion = alchemy.AlchemicalRegion(alchemical_atoms=soluteIndices, 
                                            annihilate_electrostatics=False,
                                            softcore_alpha=0.5,
                                            softcore_a=1, softcore_b=1, softcore_c=1,
                                            softcore_beta=0.0,
                                            softcore_d=1, softcore_e=1, softcore_f=1
                                           )
```

Obviously consistent_exceptions plays a role here too, but based on other discussions I've seen around this point I'm pretty sure I want to keep it as False.

I'm sorry if this is documented and I missed it.  I also know there has been a lot of tangentially-related discussion on github, but I'm having difficulty synthesizing it into something cohesive.

**Comment by jchodera** [comment]
One thing you'll need to decide is how you want to represent intra-alchemical electrostatics, especially for the fully coupled state. If you want this to be identical to the PME energy, if you simply use a Coulomb term, you are going to miss the interaction with periodic images of the alchemical system, which can be problematic for charged or even highly polar ligands. If you want to keep this as PME interactions, you will run into trouble with charged ligands. Check out the new "offset" feature in the development build of openmm 7.3, which allows you to add a charge offset to particles and exceptions you can control with a context parameter.

Let's keep this issue open to remind ourselves to more clearly explain our alchemical modifications in the Sphinx theory docs.

**Comment by JIMonroe** [comment]
Thanks!  The offset feature sounds promising, but I may wait for the official release.

>  If you want this to be identical to the PME energy, if you simply use a Coulomb term, you are going to miss the interaction with periodic images of the alchemical 

From [this forum post](https://simtk.org/plugins/phpBB/viewtopicPhpbb.php?f=161&t=4213&p=9821&start=0&view=) I was lead to believe that even if a particle-particle interaction is excluded, the reciprocal space interaction is still present as long as the charges are not zero.  Is this right?  If it works this way, you can use exceptions to turn off only direct-space intra-alchemical electrostatics without getting rid of the reciprocal space term.  I then migrate these direct-space intra-alchemical terms to exceptions in a new NonbondedForce.

From my tests, the potential energy of the alchemically-modified system with the alchemical molecule fully coupled is different than before making the alchemical modifications if I use 'direct-space' or 'coulomb'.  It _is_ the same for 'exact', but then I have to annihilate the charges, which isn't the worst thing in the world, it just means I have to do one more calculation.  I realize this is expected behavior, but in my mind it should be very clear that when you use 'direct-space' or 'coulomb' you are NOT simulating the same coupled state anymore, and are NOT using PME in the sense that one might imagine.  You do say this in the code, but thank you for clarifying here.

**Comment by andrrizzi** [comment]
> From this forum post I was lead to believe that even if a particle-particle interaction is excluded, the reciprocal space interaction is still present as long as the charges are not zero. Is this right?

I don't think this is the case. I believe OpenMM subtracts the self-term using the original charges even for pairs of atoms that are an exclusion so effectively they should not be included in the reciprocal space.

It may be possible to implement in Python a direct-space PME preserving the reciprocal space by keeping all the charges on in the `NonbondedForce` and adding a `CustomNonbondedForce` that _subtracts_ the alchemical direct-space electrostatics energy. It will be probably less efficient since you end up computing the same terms twice, but it may be more useful than what we currently have implemented.

**Comment by JIMonroe** [comment]
` I believe OpenMM subtracts the self-term using the original charges even for pairs of atoms that are an exclusion so effectively they should not be included in the reciprocal space.`

Ah, thanks a ton!  That got me on the right track and I found [this discussion](https://github.com/pandegroup/openmm/issues/1764) that really made things clear.

I don't want to spend too much more time on this since it sounds like the "offset" feature will make this discussion more or less obsolete, but I do want to sort of sum up what I've learned (please correct me if I'm still off somewhere).  The approach I outlined is not exactly what I described I want because, although it _does_ still include non-alchemical/alchemical electrostatics in the original non-bonded force, including the reciprocal part, it does not include the reciprocal part of the alchemical-alchemical interaction.  So as @jchodera mentioned, there will still be issues with charged or highly-polar ligands.  In 'direct-space' or 'coulomb' you're more honest about what you're doing and switch off _all_ reciprocal space interactions involving the alchemical molecule, leaving only the direct-space interactions which are encoded in the custom forces you create.

This has really helped my understanding, so thank you!  Hopefully it also helps write the documentation clearly (assuming the "offset" stuff doesn't completely supersede all of this).

**Comment by andrrizzi** [comment]
> turn on a CustomNonbonded force that defines regular Coulomb interactions but applies an interaction group for just alchemical-alchemical

This is something that we can implement in `AbsoluteAlchemicalFactory` to enable decoupling of electrostatics with exact PME treatment! In the gas phase, we're not interested in the interactions between periodic copies of the alchemical molecule anyway. I hope I'll be able to wrap up the offset enhancement in the next few days, but if you want this feature in `openmmtools` feel free to open a PR with the changes, and I can help you out.

> I can get it to work very efficiently as long as I keep a different context object in memory for every alchemical state (otherwise it's slow if you want dispersion corrections).

The Coulomb forces should have dispersion interactions set to `False`. For the sterics, our strategy is to disable them (with `AbsoluteAlchemicalFactory(disable_alchemical_dispersion_correction=False)`) and reweight the end points to account for them again. This allows us to use a single `Context` for everything, which means we can simulate as many states as we want without limits imposed by the GPU memory, and the loss of overlap is usually very small since the number of alchemical atoms is usually small.

**Comment by andrrizzi** [comment]
I've changed the title of this to remind us to implement decoupled electrostatics with exact PME and direct-space PME preserving the reciprocal space during the alchemical transformation.

**Comment by JIMonroe** [comment]
Thank you!  Sorry for the silence for awhile, but I will fork and take a stab at modifying alchemistry.py to include this option when I get the chance.  For now it's been easier for me to just stick with my own code, but you guys have a lot of nice features I eventually want to take advantage of.

**Comment by jchodera** [comment]
@JIMonroe : Just wanted to check in to see if you had a chance to try to make these changes, or if we should give it a try!
